File: 1325-path-with-maximum-probability/1325-path-with-maximum-probability.cpp
class Solution {
public:
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& succProb, int start_node, int end_node) {
        // makes adjacency list with destination-probability pair
        vector<vector<pair<int, double>>> adj(n);
        for (int i = 0; i < edges.size(); i++) {
            int u = edges[i][0], v = edges[i][1];
            double w = succProb[i];
            adj[u].push_back(make_pair(v, w));
            adj[v].push_back(make_pair(u, w));
        }

        // Dijkstra but:
        // 1. finding the longest "distance" so use max-heap and flip relaxation conditions
        // 2. the "distance" is actually probability
        vector<double> prob(n, 0.0);
        priority_queue<pair<double, int>> q;
        prob[start_node] = 1.0;
        q.push(make_pair(1.0, start_node));

        while (!q.empty()) {
            auto [p, u] = q.top();
            q.pop();

            if (p == prob[u]) {
                for (auto [v, w] : adj[u]) {
                    // calculates candidate probability
                    double candidate;
                    if (prob[u] == 1.0 || prob[u] == 0.0) {
                        candidate = w;
                    } else {
                        candidate = prob[u] * w;
                    }

                    // reverse relaxation condition
                    if (candidate > prob[v]) {
                        prob[v] = candidate;
                        q.push(make_pair(candidate, v));
                    }
                }
            }
        }

        // if not reachable, it will have 0.0
        return prob[end_node];
    }
};

File: 1325-path-with-maximum-probability/1325-path-with-maximum-probability.py
class Solution:
    def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start_node: int, end_node: int) -> float:
        # Build the adjacency list with weight attribute
        adj = [ [] for _ in range(n) ]
        for i in range(len(edges)):
            u, v = edges[i]
            weight = succProb[i]
            adj[u].append((v, weight))
            adj[v].append((u, weight))

        # Reverse Dijkstra
        dist = [0] * n
        q = []
        heappush(q, (1, start_node))

        while q:
            p, u = heappop(q)

            if -p <= dist[u]:  # I think == does not work bc of floating point precision?
                for v, weight in adj[u]:
                    # probability calculation instead of weight
                    if dist[u] == 0 or dist[u] == 1:
                        d = weight
                    else:
                        d = dist[u] * weight
    
                    # reversed if statement and negative push because we are finding maximum instead of shortest
                    if d > dist[v]:
                        dist[v] = d
                        heappush(q, (-d, v))
        
        return dist[end_node]